### PR TITLE
niv/CERT-5578 Third Challenge

### PR DESCRIPTION
**Pull request by nivcertora**

### CODE REVIEW

File: contracts/truster/FixedTrusterLenderPool.spec [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+*/
+
+invariant poolAllowanceAlwaysZero(address user)
+    token.allowance(currentContract, user) == 0
+    {
+        preserved with (env e)
+        {
+            require e.msg.sender != currentContract;
+        } 
+    }
```
Comment (nd-certora) [inline contracts/truster/FixedTrusterLenderPool.spec]: it should be the same spec for both version,
can we just have `execute(bytes)` as part of the dispatching options

File: contracts/truster/truster.conf [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    "link": [
+        "TrusterLenderPool:damnValuableToken=DamnValuableToken"
+    ],
+    "optimistic_loop": true,
+    "server": "production",
+    "prover_version" : "master",
+    "rule_sanity" : "basic",
+
+    "optimistic_contract_recursion": true,
+    "contract_recursion_limit": "4",
```
Comment (nd-certora) [inline contracts/truster/truster.conf:17]: either remove or add currentContract._ to be an option for the dispatcher

**Comment by nd-certora** [comment]
lets try to make this a nice example for the dispatcher list , so  try to add currentContract as well

File: contracts/truster/FixedTrusterLenderPool.spec (hunk 1 above)
Comment (nivcertora) [reply contracts/truster/FixedTrusterLenderPool.spec]: Done

File: contracts/truster/truster.conf (hunk 2 above)
Comment (nivcertora) [reply contracts/truster/truster.conf:17]: added currentContract._ to be an option for the dispatcher

File: contracts/truster/ReceiverDummy.sol [hunk 3]
```diff
@@ -0,0 +1,11 @@
+pragma solidity ^0.8.20;
+
+import "./FixedTrusterLenderPool.sol";
```
Comment (nd-certora) [inline contracts/truster/ReceiverDummy.sol]: remove

File: contracts/truster/ReceiverDummy.sol [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract ReceiverDummy is IReceiver{
```
Comment (nd-certora) [inline contracts/truster/ReceiverDummy.sol]: no need to inherit

File: contracts/truster/TrusterLenderPool.spec [hunk 5]
```diff
@@ -0,0 +1,27 @@
+
+using DamnValuableToken as token;
+
+methods{
+    function token.totalSupply() external returns (uint256) envfree;
+    function token.allowance(address, address) external returns (uint256) envfree;
+    function _.execute(bytes) external => DISPATCHER(true);
```
Comment (nd-certora) [inline contracts/truster/TrusterLenderPool.spec:7]: no need

File: contracts/truster/truster.conf [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        "contracts/truster/TrusterLenderPool.sol",
+        "contracts/DamnValuableToken.sol"
+    ],
+    "verify": "TrusterLenderPool:contracts/truster/TrusterLenderPool.spec",
+    // Link a field in the contract to another contract.
+    "link": [
+        "TrusterLenderPool:damnValuableToken=DamnValuableToken"
+    ],
+    "optimistic_loop": true,
+    "server": "production",
```
Comment (nd-certora) [inline contracts/truster/truster.conf]: remove server + prover_version

Comment (nd-certora) [reply contracts/truster/truster.conf]: add reciever

File: contracts/truster/ReceiverDummy.sol (hunk 3 above)
Comment (nivcertora) [reply contracts/truster/ReceiverDummy.sol]: Done

File: contracts/truster/ReceiverDummy.sol (hunk 4 above)
Comment (nivcertora) [reply contracts/truster/ReceiverDummy.sol]: Done

File: contracts/truster/TrusterLenderPool.spec (hunk 5 above)
Comment (nivcertora) [reply contracts/truster/TrusterLenderPool.spec:7]: Done

File: contracts/truster/truster.conf (hunk 6 above)
Comment (nivcertora) [reply contracts/truster/truster.conf]: Done

File: contracts/truster/TrusterLenderPool.spec (hunk 5 above)
Comment (nivcertora) [reply contracts/truster/TrusterLenderPool.spec:7]: Not working without it , wait for @eytan-cert  response regarding the dispatch list feature

Comment (eytan-cert) [reply contracts/truster/TrusterLenderPool.spec:7]: So, after looking at the code with @nivcertora, we saw that the call's method sighash is always resolved. The dispatch list feature (unresolved call summary) only applies when the method's sighash is unresolved (`|possible_sighashes| != 1`). I believe `Dispatcher` is the appropriate summary in this case (method is resolved, contract is not)

Comment (nd-certora) [reply contracts/truster/TrusterLenderPool.spec:7]: good to know

**Review by nd-certora — APPROVED** [review]

**Pull request merged**